python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,576 @@
+        keep_mask = count_df['token'].isin(keep_idx)
+        count_df = count_df.loc[count_df.index[keep_mask]]
+        count_df['token'] = count_df['token'].astype(cp.int32)
+
+        if keep_num == 1 and isinstance(count_df, cudf.Series):
+            # Workaround for cudf bug where df.loc[[x]] returns a Series
+            # instead of a DataFrame. This if statement and its content can be
+            # safely removed once below issue is fixed.
+            # See https://github.com/rapidsai/cudf/issues/5330
+            count_df = cudf.DataFrame({x: count_df[x] for x in count_df.index})
+
+        make_monotonic(count_df['token'])
+
+        return count_df
+
+    def _preprocess(self, raw_documents):
+        preprocess = self.build_preprocessor()
+        docs = raw_documents._column.nvstrings
+        return Series(preprocess(docs))
+
+    def fit(self, raw_documents):
+        """Build a vocabulary of all tokens in the raw documents.
+
+       Parameters
+       ----------
+       raw_documents : cudf.Series
+           A Series of string documents
+
+       Returns
+       -------
+       self
+       """
+        self.fit_transform(raw_documents)
+        return self
+
+    def fit_transform(self, raw_documents):
+        """Build the vocabulary and return document-term matrix.
+
+        Equivalent to .fit(X).transform(X) but preprocess X only once.
+
+        Parameters
+        ----------
+        raw_documents : cudf.Series
+           A Series of string documents
+
+        Returns
+        -------
+        X : array of shape (n_samples, n_features)
+            Document-term matrix.
+        """
+        self._warn_for_unused_params()
+        self._validate_params()
+
+        self._fixed_vocabulary = self.vocabulary is not None
+
+        docs = self._preprocess(raw_documents)
+        n_doc = len(docs)
+
+        tokenized_df = self._create_tokenized_df(docs)
+
+        if self._fixed_vocabulary:
+            self.vocabulary_ = self.vocabulary
+        else:
+            self.vocabulary_ = tokenized_df["token"].unique()
+
+        count_df = self._count_vocab(tokenized_df)
+
+        if not self._fixed_vocabulary:
+            max_doc_count = (self.max_df
+                             if isinstance(self.max_df, numbers.Integral)
+                             else self.max_df * n_doc)
+            min_doc_count = (self.min_df
+                             if isinstance(self.min_df, numbers.Integral)
+                             else self.min_df * n_doc)
+            if max_doc_count < min_doc_count:
+                raise ValueError(
+                    "max_df corresponds to < documents than min_df")
+            count_df = self._limit_features(count_df, self.vocabulary_,
+                                            max_doc_count,
+                                            min_doc_count,
+                                            self.max_features)
+
+        empty_doc_ids = self._compute_empty_doc_ids(count_df, n_doc)
```

Comment:
Should the second half of this just call `transform`? Or are there perf gains from the repetition here?

Reply:
There is no perf gain but the second half of fit_transform is not exactly the same as `transform`. `fit_transform` does some extra work in between the calls to `_create_tokenized_df()`, `_count_vocab()` and `_compute_empty_doc_ids` which `transform` doesn't do.